### ISSUE TITLE
access-order of fields is not documented correctly -- SIMICS-19021

### ISSUE THREAD

**Issue by lwaern-intel** [intel/device-modeling-language#77]
The documentation for read_register states that fields are read lsb first.
The documentation for write_register states that fields are written msb first.
The actual implementation is:

For read:
1. fields without read-template are retrieved all at once
2. fields with read-template are read lsb first

For write:
1. fields with write-template are written lsb first
2. fields without write-template are written all at once

Either documentation or implementation needs to be updated.

**ecarsten added a comment - 10/Mar/22 2:01 PM:**
Doc should be updated so read and write use the same direction. Implementation should be updated accordingly (looks easy).

**ecarsten added a comment - 11/Mar/22 7:54 PM:**
The documentation for read_register states that fields are read lsb first.

I would interpret the documentation for read as "msb first":

```
The implementation invokes the `read_field` method of
all sub-fields at least partially covered by `enabled_bytes`, in order
from most to least significant bit.  
```
In any case, this is inconsistent with the implementation, which accesses LSB first. Test of current behaviour:

```
--- a/test/1.4/lib/T_field.dml
+++ b/test/1.4/lib/T_field.dml
@@ -37,4 +37,14 @@ bank b {
     register r4 size 4 @ 12 {
         field f @ [31:0] is use_get_val;
     }
+
+    register r5 size 4 @ 16 {
+        saved uint64 written;
+        field a[i < 4] @ [i * 8 + 7 : i * 8] is write {
+            method write(uint64 val) {
+                written <<= 8;
+                written[7:0] = val;
+            }
+        }
+    }
 }
diff --git a/test/1.4/lib/T_field.py b/test/1.4/lib/T_field.py
index 9496f1d9..3233107e 100644
--- a/test/1.4/lib/T_field.py
+++ b/test/1.4/lib/T_field.py
@@ -7,3 +7,7 @@ for name in ['r1', 'r2', 'r3', 'r4']:
     setattr(obj, 'b_' + name, 4)
     # set adds 1, get adds 2
     stest.expect_equal(getattr(obj, 'b_' + name), 7)
+
+dev_util.Register_LE(obj.bank.b, 16).write(0xaabbccdd)
+# lsb written first
+stest.expect_equal(obj.bank.b.r5_written, 0xddccbbaa)
```
The 1.2 behaviour is to access MSB first, as documented, so either we have to switch existing 1.4 devs from LSB-first to MSB-first, or we have to migrate every remaining 1.2 dev from MSB-first to LSB-first while migrating to 1.4 (possibly with extra pain from supporting both orders in parallel during a migration window). Both seem like similar amounts of work. Any opinions on which one is preferable?

I recall from ancient times that the 1.2 order (MSB first) happened to be practical more often than it was impractical. OTOH, most regs use little-endian byte order which makes "LSB first" feel more natural (it means that the field with lowest address is written first, if different fields cover different bytes).

**Comment by mandolaerik** [comment]
There are three options:

1. MSB first
2. LSB first
3. Lowest address first, i.e. MSB first or LSB first depending on the bank's byte_order

Lowest address first would be a natural choice (it is consistent with access order of accesses that span multiple registers), but it's more complicated both to implement and explain. And we have seen use cases where a bank is instantiated twice with different byte orders; having different access orders for those banks would be destructive.
So, best to pick 1 or 2 to keep things simple and clear. We already do LSB first and it is the natural choice for little-endian byte order which is the default, so let's just test and document that.

**Comment by lwaern-intel** [comment]
Fixed by #79
